### ISSUE TITLE
BEP 52: The BitTorrent Protocol Specification v2 (Plans or roadmap)?

### ISSUE THREAD

**Issue by jeffward01** [alanmcgovern/monotorrent#480]
# Introduction

This is not a bug or 'issue', just open dialog and conversation

###  Question 1: Is there an existing roadmap (ordered todo list) or plans to implement: http://bittorrent.org/beps/bep_0052.html 

It will require a major re-write of some areas and workflows.

If acceptable, i do not mind working on this.

It seems that we need a way to either:

* **A.)** Create C# bindings for: http://libtorrent.org/

or

* **B.)**  Make efforts as a community (not just from the super-human programmer @alanmcgovern) to implement the rest of the BEP features which is present in LibTorrent.

Currently, if we use any software which does not implement all of the BEP's which [LibTorrent](libtorrent.org) implements in C++, we are compromising features and performance for sake of programming language choice.

## Side Notes:

* **1.)**  Does anyone know how difficult it is to create C# bindings for libtorrent using: [CppSharp](https://github.com/mono/CppSharp) or other microsoft toolings .

* **2.)** @alanmcgovern - do you have a roadmap or sequence of BEP features which you would like implements, or is it pick one from the hat and implement

* **3.)** I do not think @alanmcgovern gets enough credit. YOU ALONE ARE POWERING THE DOTNET COMMUNITY FOR BITTORRENT TECHNOLOGY! I hope you realize how amazing and awesome this is!!!!!!!!!!!!!!!!!!!!!!

**Comment by BoyFaceGirl** [comment]
> # Introduction
> This is not a bug or 'issue', just open dialog and conversation
> 
> ## Questions
> ### Question 1: Is there an existing roadmap (ordered todo list) or plans to implement: http://bittorrent.org/beps/bep_0052.html
> It will require a major re-write of some areas and workflows.
> 
> If acceptable, i do not mind working on this.
> 
> It seems that we need a way to either:
> 
> * **A.)** Create C# bindings for: http://libtorrent.org/
> 
> or
> 
> * **B.)**  Make efforts as a community (not just from the super-human programmer @alanmcgovern) to implement the rest of the BEP features which is present in LibTorrent.
> 
> Currently, if we use any software which does not implement all of the BEP's which [LibTorrent](libtorrent.org) implements in C++, we are compromising features and performance for sake of programming language choice.
> 
> ### Question 2: Is there a community MonoTorrent discord or discussion channel?
> ## Side Notes:
> * **1.)**  Does anyone know how difficult it is to create C# bindings for libtorrent using: [CppSharp](https://github.com/mono/CppSharp) or other microsoft toolings .
> * **2.)** @alanmcgovern - do you have a roadmap or sequence of BEP features which you would like implements, or is it pick one from the hat and implement
> * **3.)** I do not think @alanmcgovern gets enough credit. YOU ALONE ARE POWERING THE DOTNET COMMUNITY FOR BITTORRENT TECHNOLOGY! I hope you realize how amazing and awesome this is!!!!!!!!!!!!!!!!!!!!!!


I don't recommend Create C# bindings for: http://libtorrent.org/，It will lose the fun of pure c# implementation. The c# syntax is simple enough, which can make it easier for more people to learn the biotorrent principle. If binding  for: http://libtorrent.org/，Why not bind directly http://aria2.github.io/ ？ So the energy supply is more perfect and easier?

**Comment by alanmcgovern** [comment]
> A.) Create C# bindings for: http://libtorrent.org/

If you want to go the route of binding libtorrent, then resuming this work would likely be the best route - https://github.com/arvidn/libtorrent/issues/116 . The result would not be quite as cross-platform/easy to consume, but should be fine!

> B.) Make efforts as a community to implement the rest of the BEP 

I don't think it's too difficult to add support for bittorrent v2. The most complex part is handling the changes to how the hashes are generated/used. Once that's done, I *think* the remainder of the work could be completed in a few days.

In slightly more detail:  'MetadataMode' exists today, and is used to download .torrent metadata from other peers before initiating the download. It would be relatively simple to add a similar mode which retrieves all required hashes from other peers before beginning the download. This could later be optimised so hashes are retrieved in parallel with the actual downloading process.

Also, MonoTorrent has the concept of treating pieces as undownloadable, and then dynamically marking them as 'downloadable' later on. i.e. you can safely download a torrent while hashchecking it. This would mesh well with this capability.

> do you have a roadmap or sequence of BEP features which you would like implements, or is it pick one from the hat and implement

No real preference. It just depends on what people want, and what I'm interested in implementing!

> I do not think @alanmcgovern gets enough credit

Thanks :p 

> If acceptable, i do not mind working on this.

If you're interested in working on this, a great first step would be to upgrade the `Torrent` class to be able to parse/load V2 torrents. Then update `TorrentCreator` so it can generate v2 torrents. Once that's in place, I can sketch out the next phase of work to begin downloading torrents using the v2 hashes.

**Comment by alanmcgovern** [comment]
https://github.com/alanmcgovern/monotorrent/pull/481

This is a really crappy spike to show one potential way to parse v2 torrents, and support hybrid v1/v2 torrents if the approach were tweaked to support loading both the v2 layers and v1 hashes into the ITorrentFileHashProvider interface thing I threw together. 

If you're interested in working more on this and want some guidance, let me know!

**Comment by alanmcgovern** [comment]
Ah, I remember why I looked at this before and decided not to do it.

The 'spec' is so poorly written it's not possible to implement without pulling in other sources of information. How, precisely, are the three new messages (`hash request`, `hashes`, hash reject`) constructed/encoded and parsed/decoded? 

How, precisely, are the 'proof layers' intended to be used.

How, precisely, do you 'omit' a proof layer? What does it mean when it says 'The omitted layers are still counted towards the requested proof layers'? If a proof layer is omitted then why is it being counted?

What exactly happens if 'proof layers' is set to 0, 1, 2, 3... etc?

Regardless of the semantics of how these things are supposed to be used, the basics of how the messages are converted to/from a byte[] to be sent between peers are ambiguous at best.

**Comment by alanmcgovern** [comment]
Closing this due to lack of engagement. 

BEP52 is not particularly impactful in terms of improvements, and is a fair chunk of work all-in. I may eventually finish it, but it's pretty low priority.
